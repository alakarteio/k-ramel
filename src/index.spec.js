/* eslint-env jest */
import { createStore, simpleObject, keyValue, applyMiddleware } from './index'

describe('k-simple-state', () => {
  const simpleTests = (getStore) => {
    it('should initialized', () => {
      const store = getStore()
      expect(store.getState()).toMatchSnapshot()
    })

    it('should instanciate a keyValue reducer', () => {
      const store = getStore()

      // mutation
      store.data.todos.add({ id: 2, label: 'find a name' })

      // selection
      const todo = store.data.todos.get(2)

      expect({
        state: store.getStore().getState(),
        todo,
      }).toMatchSnapshot()
    })

    it('should instanciate a simpleObject reducer', () => {
      const store = getStore()

      // mutation
      store.ui.screens.newTodo.set('finish tests')

      // selection
      const label = store.ui.screens.newTodo.get()

      expect({
        state: store.getState(),
        label,
      }).toMatchSnapshot()
    })
  }

  describe('with plain object', () => {
    const getNewStore = () => createStore({
      data: {
        todos: { type: 'keyValue', key: 'id' },
      },
      ui: {
        screens: {
          newTodo: { type: 'simpleObject' },
        },
      },
    })

    simpleTests(getNewStore)
  })

  describe('with helpers', () => {
    const getNewStore = () => createStore({
      data: {
        todos: keyValue({ key: 'id' }),
      },
      ui: {
        screens: {
          newTodo: simpleObject(),
        },
      },
    })

    simpleTests(getNewStore)
  })

  describe('with raw reducers', () => {
    const getNewStore = () => createStore({
      data: {
        todos: keyValue({ key: 'id' }),
      },
      ui: {
        config: (state = 'defaultState', action) => {
          switch (action.type) {
            case 'SET_CONFIG': return action.payload
            default: return state
          }
        },
        screens: {
          newTodo: simpleObject(),
        },
      },
    })

    simpleTests(getNewStore)

    it('should dispatch action to the raw reducer', () => {
      const store = getNewStore()

      // mutation
      store.dispatch({ type: 'SET_CONFIG', payload: 'new config' })

      // selection
      const { config } = store.getState().ui

      expect({
        state: store.getState(),
        config,
      }).toMatchSnapshot()
    })
  })

  describe('without custom options', () => {
    const getNewStore = options => createStore({
      data: {
        todos: { type: 'keyValue', key: 'id' },
      },
      ui: {
        screens: {
          newTodo: { type: 'simpleObject' },
        },
      },
    }, options)

    describe('middlewares', () => {
      it('should call middlewares', () => {
        const spy = jest.fn()
        const middleware = store => next => (action) => {
          spy({ state: store.getState(), action })
          next(action)
        }
        const store = getNewStore({ enhancer: applyMiddleware(middleware) })
        store.ui.screens.newTodo.set('new')

        expect({
          state: store.getState(),
          spy: spy.mock.calls,
        }).toMatchSnapshot()
      })
    })

    describe('init', () => {
      it('should init the state', () => {
        const store = getNewStore({
          init: {
            ui: {
              screens: {
                newTodo: 'initial value!',
              },
            },
          },
        })

        expect({
          state: store.getState(),
        }).toMatchSnapshot()
      })
    })

    describe('hide redux', () => {
      it('should not mutate the state', () => {
        const store = getNewStore({ hideRedux: false })
        store.data.todos.add({ id: '3', label: 'hide-redux' })
        expect({
          state: store.getState(),
        }).toMatchSnapshot()
      })

      it('should dispatch action', () => {
        const store = getNewStore({ hideRedux: false })
        store.dispatch(store.data.todos.add({ id: '3', label: 'hide-redux' }))
        expect({
          state: store.getState(),
          todo: store.data.todos.get(3)(store.getState()),
        }).toMatchSnapshot()
      })
    })
  })
})
