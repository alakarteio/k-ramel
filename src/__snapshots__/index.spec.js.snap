// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bugs should works with first level reducers 1`] = `
Object {
  "label": "yeah it works",
  "state": Object {
    "label": "yeah it works",
  },
}
`;

exports[`k-simple-state with helpers should initialized 1`] = `
Object {
  "data": Object {
    "todos": Object {
      "array": Array [],
      "data": Object {},
      "initialized": false,
      "keys": Array [],
    },
  },
  "ui": Object {
    "screens": Object {
      "newTodo": Object {},
    },
  },
}
`;

exports[`k-simple-state with helpers should instanciate a keyValue reducer 1`] = `
Object {
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [
          Object {
            "id": 2,
            "label": "find a name",
          },
        ],
        "data": Object {
          "2": Object {
            "id": 2,
            "label": "find a name",
          },
        },
        "initialized": true,
        "keys": Array [
          2,
        ],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": Object {},
      },
    },
  },
  "todo": Object {
    "id": 2,
    "label": "find a name",
  },
}
`;

exports[`k-simple-state with helpers should instanciate a simpleObject reducer 1`] = `
Object {
  "label": "finish tests",
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": "finish tests",
      },
    },
  },
}
`;

exports[`k-simple-state with plain object should initialized 1`] = `
Object {
  "data": Object {
    "todos": Object {
      "array": Array [],
      "data": Object {},
      "initialized": false,
      "keys": Array [],
    },
  },
  "ui": Object {
    "screens": Object {
      "newTodo": Object {},
    },
  },
}
`;

exports[`k-simple-state with plain object should instanciate a keyValue reducer 1`] = `
Object {
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [
          Object {
            "id": 2,
            "label": "find a name",
          },
        ],
        "data": Object {
          "2": Object {
            "id": 2,
            "label": "find a name",
          },
        },
        "initialized": true,
        "keys": Array [
          2,
        ],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": Object {},
      },
    },
  },
  "todo": Object {
    "id": 2,
    "label": "find a name",
  },
}
`;

exports[`k-simple-state with plain object should instanciate a simpleObject reducer 1`] = `
Object {
  "label": "finish tests",
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": "finish tests",
      },
    },
  },
}
`;

exports[`k-simple-state with raw reducers should dispatch action to the raw reducer 1`] = `
Object {
  "config": "new config",
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "config": "new config",
      "screens": Object {
        "newTodo": Object {},
      },
    },
  },
}
`;

exports[`k-simple-state with raw reducers should initialized 1`] = `
Object {
  "data": Object {
    "todos": Object {
      "array": Array [],
      "data": Object {},
      "initialized": false,
      "keys": Array [],
    },
  },
  "ui": Object {
    "config": "defaultState",
    "screens": Object {
      "newTodo": Object {},
    },
  },
}
`;

exports[`k-simple-state with raw reducers should instanciate a keyValue reducer 1`] = `
Object {
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [
          Object {
            "id": 2,
            "label": "find a name",
          },
        ],
        "data": Object {
          "2": Object {
            "id": 2,
            "label": "find a name",
          },
        },
        "initialized": true,
        "keys": Array [
          2,
        ],
      },
    },
    "ui": Object {
      "config": "defaultState",
      "screens": Object {
        "newTodo": Object {},
      },
    },
  },
  "todo": Object {
    "id": 2,
    "label": "find a name",
  },
}
`;

exports[`k-simple-state with raw reducers should instanciate a simpleObject reducer 1`] = `
Object {
  "label": "finish tests",
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "config": "defaultState",
      "screens": Object {
        "newTodo": "finish tests",
      },
    },
  },
}
`;

exports[`k-simple-state without custom options hide redux should dispatch action 1`] = `
Object {
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [
          Object {
            "id": "3",
            "label": "hide-redux",
          },
        ],
        "data": Object {
          "3": Object {
            "id": "3",
            "label": "hide-redux",
          },
        },
        "initialized": true,
        "keys": Array [
          "3",
        ],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": Object {},
      },
    },
  },
  "todo": Object {
    "id": "3",
    "label": "hide-redux",
  },
}
`;

exports[`k-simple-state without custom options hide redux should not mutate the state 1`] = `
Object {
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": Object {},
      },
    },
  },
}
`;

exports[`k-simple-state without custom options init should init the state 1`] = `
Object {
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": "initial value!",
      },
    },
  },
}
`;

exports[`k-simple-state without custom options middlewares should call middlewares 1`] = `
Object {
  "spy": Array [
    Array [
      Object {
        "action": Object {
          "payload": "new",
          "type": "@@krf/SET_NEWTODO",
        },
        "state": Object {
          "data": Object {
            "todos": Object {
              "array": Array [],
              "data": Object {},
              "initialized": false,
              "keys": Array [],
            },
          },
          "ui": Object {
            "screens": Object {
              "newTodo": Object {},
            },
          },
        },
      },
    ],
  ],
  "state": Object {
    "data": Object {
      "todos": Object {
        "array": Array [],
        "data": Object {},
        "initialized": false,
        "keys": Array [],
      },
    },
    "ui": Object {
      "screens": Object {
        "newTodo": "new",
      },
    },
  },
}
`;

exports[`listen middleware should catch action -function- 1`] = `
Array [
  Object {
    "payload": "this is dispatched !",
    "type": "@@krf/SET_CONFIG",
  },
  Object {
    "config": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
]
`;

exports[`listen middleware should catch action -regexp- 1`] = `
Array [
  Object {
    "payload": "this is dispatched !",
    "type": "@@krf/SET_CONFIG",
  },
  Object {
    "config": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
]
`;

exports[`listen middleware should catch action decorated by reaction hof 1`] = `
Array [
  Object {
    "payload": "this is dispatched !",
    "type": "@@krf/SET_CONFIG",
  },
  Object {
    "config": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
]
`;

exports[`listen middleware should catch action decorated by reaction hof, DSL way 1`] = `
Array [
  Object {
    "payload": "this is dispatched !",
    "type": "@@krf/SET_CONFIG",
  },
  Object {
    "config": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
]
`;

exports[`listen middleware should catch action decorated by reaction hof, DSL way, with multiple keys 1`] = `
Array [
  Object {
    "payload": "this is dispatched !",
    "type": "@@krf/SET_CONFIG",
  },
  Object {
    "config": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
]
`;

exports[`listen middleware should catch and dispatch a new action 1`] = `
Object {
  "state": Object {
    "config": "this is dispatched !",
    "saved": "SET_CONFIG is triggered :)",
  },
}
`;

exports[`listen middleware should still dispatch events 1`] = `
Object {
  "config": "this is dispatched !",
  "state": Object {
    "config": "this is dispatched !",
  },
}
`;
